# Exploratory Data Analysis – World Happiness Report

**Course:** Programming for Artificial Intelligence and Data Science (P4AI-DS)
**Assignment:** 1 - Exploratory Data Analysis (Tabular Data)
**Dataset:** World Happiness Report 2019
**Source:** [Kaggle - World Happiness Report](https://www.kaggle.com/datasets/unsdsn/world-happiness)

---
> **Disclaimer:** Môi trường hiển thị đồ thị được lập trình lại 100% bằng kiến trúc `plotly.graph_objects`.
---

In [52]:
import pandas as pd
import plotly.express as px
import plotly.graph_objects as go
from plotly.subplots import make_subplots
import warnings
warnings.filterwarnings('ignore')

# Load data
df = pd.read_csv('2019.csv')
print("✅ Libraries loaded successfully!")

✅ Libraries loaded successfully!


## 1. Dataset Overview

In [53]:
print(f"Dataset overview")
print(f"{len(df):,}")
print("Training Samples")
print(f"{len(df.columns)}")
print("Total Features")

numerical_features = df.select_dtypes(include='number').columns.tolist()
categorical_features = df.select_dtypes(exclude='number').columns.tolist()

# Feature Engineering (Tạo Feature Categorical mới từ Numerical GDP để đánh giá phân phối tự nhiên)
df['GDP_Level'] = pd.cut(df['GDP per capita'], bins=3, labels=['Low', 'Medium', 'High'])
categorical_features.append('GDP_Level')

print(f"{len(numerical_features)}")
print("Numerical Features")
print(f"{len(categorical_features)}")
print("Categorical Features")
print(f"Target Variable: Score")

print("\n📋 Feature Descriptions")
print(f"{'Feature Name':<28} | {'Type':<18} | {'Description'}")
print("-"*80)
print(f"{'Overall rank':<28} | {'Numerical':<18} | {'Rank of the country'}")
print(f"{'Country or region':<28} | {'Categorical':<18} | {'Name of the country'}")
print(f"{'Score':<28} | {'Numerical (Target)':<18} | {'Happiness score'}")
print(f"{'GDP per capita':<28} | {'Numerical':<18} | {'Economic production per person'}")
print(f"{'Social support':<28} | {'Numerical':<18} | {'Perceived social support networks'}")
print(f"{'Healthy life expectancy':<28} | {'Numerical':<18} | {'Expectancy of healthy life'}")
print(f"{'Freedom to make life choices':<28} | {'Numerical':<18} | {'Perceived freedom'}")
print(f"{'Generosity':<28} | {'Numerical':<18} | {'Generosity of the country'}")
print(f"{'Perceptions of corruption':<28} | {'Numerical':<18} | {'Perceived corruption'}")
print(f"{'GDP_Level':<28} | {'Categorical (New)':<18} | {'Engineered feature (Low, Medium, High)'}")

Dataset overview
156
Training Samples
9
Total Features
8
Numerical Features
2
Categorical Features
Target Variable: Score

📋 Feature Descriptions
Feature Name                 | Type               | Description
--------------------------------------------------------------------------------
Overall rank                 | Numerical          | Rank of the country
Country or region            | Categorical        | Name of the country
Score                        | Numerical (Target) | Happiness score
GDP per capita               | Numerical          | Economic production per person
Social support               | Numerical          | Perceived social support networks
Healthy life expectancy      | Numerical          | Expectancy of healthy life
Freedom to make life choices | Numerical          | Perceived freedom
Generosity                   | Numerical          | Generosity of the country
Perceptions of corruption    | Numerical          | Perceived corruption
GDP_Level                   

## 2. Missing Values Analysis

In [54]:
# Calculate missing values
missing_counts = df.isnull().sum()
missing_pct = (df.isnull().sum() / len(df) * 100)

# Filter columns with missing values
missing_df = pd.DataFrame({
    'column': missing_counts[missing_counts > 0].index,
    'count': missing_counts[missing_counts > 0].values,
    'percentage': missing_pct[missing_counts > 0].values
}).sort_values('percentage', ascending=False)

if len(missing_df) > 0:
    print("Missing Values Summary:")
    print(missing_df)
    
    # Create bar chart
    fig = go.Figure(data=[go.Bar(
        x=missing_df['column'],
        y=missing_df['percentage'],
        marker_color='#f093fb',
        text=[f"{p:.2f}%" for p in missing_df['percentage']],
        textposition='outside'
    )])

    fig.update_layout(
        title='Missing Values by Feature (%)',
        xaxis_title='Features',
        yaxis_title='Missing Percentage',
        width=800,
        height=400
    )

    fig.show()
    
    # Imputation recommendations
    print("\n💡 Recommendations:")
    for _, row in missing_df.iterrows():
        col, pct = row['column'], row['percentage']
        if pct > 50:
            print(f"  {col}: {pct:.1f}% missing → Consider dropping")
        elif pct > 5:
            print(f"  {col}: {pct:.1f}% missing → Use advanced imputation (KNN, iterative)")
        else:
            print(f"  {col}: {pct:.1f}% missing → Simple imputation (mode/median)")
else:
    print("✅ Dataset is clean. No missing values reported.")

✅ Dataset is clean. No missing values reported.


## 3. Numerical Features Distribution

In [55]:
# Select numerical features for deep dive
features = ['GDP per capita', 'Social support', 'Healthy life expectancy', 'Freedom to make life choices', 'Generosity']

for feature in features:
    # Create histogram
    fig = go.Figure()

    fig.add_trace(go.Histogram(
        x=df[feature],
        nbinsx=30,
        name='Distribution',
        marker_color='#667eea',
        opacity=0.7
    ))

    fig.update_layout(
        title=f'{feature} Distribution',
        xaxis_title=feature,
        yaxis_title='Count',
        width=800,
        height=350,
        showlegend=True
    )

    fig.show()

    # Statistics
    print(f"{feature} Statistics:")
    print(f"  Mean: {df[feature].mean():.2f}")
    print(f"  Median: {df[feature].median():.2f}")
    print(f"  Std: {df[feature].std():.2f}")
    print(f"  Min: {df[feature].min():.2f}")
    print(f"  Max: {df[feature].max():.2f}")
    print("="*40)

GDP per capita Statistics:
  Mean: 0.91
  Median: 0.96
  Std: 0.40
  Min: 0.00
  Max: 1.68


Social support Statistics:
  Mean: 1.21
  Median: 1.27
  Std: 0.30
  Min: 0.00
  Max: 1.62


Healthy life expectancy Statistics:
  Mean: 0.73
  Median: 0.79
  Std: 0.24
  Min: 0.00
  Max: 1.14


Freedom to make life choices Statistics:
  Mean: 0.39
  Median: 0.42
  Std: 0.14
  Min: 0.00
  Max: 0.63


Generosity Statistics:
  Mean: 0.18
  Median: 0.18
  Std: 0.10
  Min: 0.00
  Max: 0.57


## 4. Categorical Features Distribution

In [56]:
# Select a categorical feature
feature = 'GDP_Level'

# Get value counts
value_counts = df[feature].value_counts()

# Create bar chart
fig = go.Figure(data=[go.Bar(
    x=value_counts.index,
    y=value_counts.values,
    marker_color='#4facfe',
    text=value_counts.values,
    textposition='outside'
)])

fig.update_layout(
    title=f'{feature.replace("_", " ").title()} Distribution',
    xaxis_title=feature.replace('_', ' ').title(),
    yaxis_title='Count',
    width=900,
    height=500,
    xaxis={'tickangle': 0}
)

fig.show()

# Print statistics
print(f"\n{feature} Statistics:")
print(f"  Unique values: {df[feature].nunique()}")
print(f"  Missing: {df[feature].isnull().sum()} ({df[feature].isnull().sum()/len(df)*100:.2f}%)\n")
print(f"Top categories:")
for cat, count in value_counts.head(10).items():
    print(f"  {cat}: {count:,} ({count/len(df)*100:.2f}%)")


GDP_Level Statistics:
  Unique values: 3
  Missing: 0 (0.00%)

Top categories:
  Medium: 68 (43.59%)
  High: 54 (34.62%)
  Low: 34 (21.79%)


## 5. Target Variable Distribution

In [57]:
# Target distribution (Happiness Score is continuous numerical)
target = 'Score'

fig = go.Figure()

fig.add_trace(go.Histogram(
    x=df[target],
    nbinsx=20,
    name='Distribution',
    marker_color='#f093fb',
    opacity=0.8
))

fig.update_layout(
    title='Happiness Score Distribution (Target Variable)',
    xaxis_title=target,
    yaxis_title='Count',
    showlegend=True,
    width=800,
    height=400
)

fig.show()

# Print statistics
print("Target Distribution Statistics:")
print(f"  Mean: {df[target].mean():.2f}")
print(f"  Median: {df[target].median():.2f}")
print(f"  Min/Max: [{df[target].min():.2f}, {df[target].max():.2f}]")

Target Distribution Statistics:
  Mean: 5.41
  Median: 5.38
  Min/Max: [2.85, 7.77]


## 6. Correlation Analysis

In [58]:
# Select numerical features
numerical_cols = ['Score', 'GDP per capita', 'Social support', 'Healthy life expectancy', 
                  'Freedom to make life choices', 'Generosity', 'Perceptions of corruption']

# Compute correlation matrix
corr_matrix = df[numerical_cols].corr()

# Create heatmap
fig = go.Figure(data=go.Heatmap(
    z=corr_matrix.values,
    x=corr_matrix.columns,
    y=corr_matrix.columns,
    colorscale='RdBu_r',
    zmid=0,
    text=corr_matrix.values,
    texttemplate='%{text:.2f}',
    textfont={"size": 10},
    colorbar=dict(title="Correlation")
))

fig.update_layout(
    title='Correlation Matrix - Numerical Features',
    width=800,
    height=700,
    xaxis={'tickangle': 45}
)

fig.show()

# Print high correlations
print("\nHigh Correlations (|r| > 0.5):")
for i in range(len(numerical_cols)):
    for j in range(i+1, len(numerical_cols)):
        corr = corr_matrix.iloc[i, j]
        if abs(corr) > 0.5:
            print(f"  {numerical_cols[i]} ↔ {numerical_cols[j]}: r = {corr:.3f}")


High Correlations (|r| > 0.5):
  Score ↔ GDP per capita: r = 0.794
  Score ↔ Social support: r = 0.777
  Score ↔ Healthy life expectancy: r = 0.780
  Score ↔ Freedom to make life choices: r = 0.567
  GDP per capita ↔ Social support: r = 0.755
  GDP per capita ↔ Healthy life expectancy: r = 0.835
  Social support ↔ Healthy life expectancy: r = 0.719


## 7. Outlier Detection (IQR Method)

In [59]:
# Select features with potential outliers
features = ['Score', 'GDP per capita', 'Social support', 'Perceptions of corruption', 'Generosity']

# Create box plots
fig = make_subplots(
    rows=1, cols=len(features),
    subplot_titles=[f.replace('_', ' ').title() for f in features]
)

colors = ['#667eea', '#f093fb', '#4facfe', '#43e97b', '#f9bc2c']

for idx, (feature, color) in enumerate(zip(features, colors), 1):
    fig.add_trace(
        go.Box(y=df[feature], name=feature, marker_color=color, boxmean='sd'),
        row=1, col=idx
    )

fig.update_layout(
    title_text='Outlier Detection - Box Plots',
    showlegend=False,
    height=450,
    width=1200
)

fig.show()

# IQR method
print("\n--- IQR Outlier Detection ---")
for feature in features:
    Q1 = df[feature].quantile(0.25)
    Q3 = df[feature].quantile(0.75)
    IQR = Q3 - Q1
    
    lower_bound = Q1 - 1.5 * IQR
    upper_bound = Q3 + 1.5 * IQR
    
    outliers = df[(df[feature] < lower_bound) | (df[feature] > upper_bound)]
    
    if len(outliers) > 0:
        print(f"\n{feature}:")
        print(f"  Q1: {Q1:.2f}, Q3: {Q3:.2f}, IQR: {IQR:.2f}")
        print(f"  Bounds: [{lower_bound:.2f}, {upper_bound:.2f}]")
        print(f"  Outliers: {len(outliers)} ({len(outliers)/len(df)*100:.2f}%)")
        print(f"  Outlier range: [{outliers[feature].min():.2f}, {outliers[feature].max():.2f}]")


--- IQR Outlier Detection ---

Social support:
  Q1: 1.06, Q3: 1.45, IQR: 0.40
  Bounds: [0.46, 2.05]
  Outliers: 4 (2.56%)
  Outlier range: [0.00, 0.45]

Perceptions of corruption:
  Q1: 0.05, Q3: 0.14, IQR: 0.09
  Bounds: [-0.09, 0.28]
  Outliers: 14 (8.97%)
  Outlier range: [0.29, 0.45]

Generosity:
  Q1: 0.11, Q3: 0.25, IQR: 0.14
  Bounds: [-0.10, 0.46]
  Outliers: 2 (1.28%)
  Outlier range: [0.50, 0.57]


## 8. Target vs Categorical Features

In [60]:
# Analyze target by categorical feature
feature = 'GDP_Level'  

fig = go.Figure()

for category in ['High', 'Medium', 'Low']:
    fig.add_trace(go.Box(
        y=df[df[feature] == category]['Score'],
        name=category,
        marker_color='#f093fb' if category == 'High' else ('#667eea' if category == 'Medium' else '#43e97b'),
        boxmean='sd'
    ))

fig.update_layout(
    title=f'Target Variable (Score) Distribution by {feature}',
    xaxis_title=feature,
    yaxis_title='Score',
    width=900,
    height=500
)

fig.show()

# Print insights
print(f"\n{feature} Statistics related to Target (Score):")
grouped = df.groupby(feature)['Score'].agg(['mean', 'median', 'std', 'max', 'min']).sort_values('mean', ascending=False)
print(grouped.round(2))


GDP_Level Statistics related to Target (Score):
           mean  median   std   max   min
GDP_Level                                
High       6.47    6.34  0.68  7.77  5.29
Medium     5.19    5.20  0.74  7.17  3.46
Low        4.16    4.32  0.65  5.47  2.85


## 9. Sample Data Rows

In [61]:
print("--- First 3 Rows ---")
display(df.head(3))

print("\n--- Random 3 Sample Rows ---")
display(df.sample(3, random_state=42))

print("\n--- Last 3 Rows ---")
display(df.tail(3))

--- First 3 Rows ---


,Overall rank,Country or region,Score,GDP per capita,Social support,Healthy life expectancy,Freedom to make life choices,Generosity,Perceptions of corruption,GDP_Level
0,1,Finland,7.769,1.340,1.587,0.986,0.596,0.153,0.393,High
1,2,Denmark,7.600,1.383,1.573,0.996,0.592,0.252,0.410,High
2,3,Norway,7.554,1.488,1.582,1.028,0.603,0.271,0.341,High



--- Random 3 Sample Rows ---


,Overall rank,Country or region,Score,GDP per capita,Social support,Healthy life expectancy,Freedom to make life choices,Generosity,Perceptions of corruption,GDP_Level
96,97,Bulgaria,5.011,1.092,1.513,0.815,0.311,0.081,0.004,Medium
69,70,Serbia,5.603,1.004,1.383,0.854,0.282,0.137,0.039,Medium
82,83,Mongolia,5.285,0.948,1.531,0.667,0.317,0.235,0.038,Medium



--- Last 3 Rows ---


,Overall rank,Country or region,Score,GDP per capita,Social support,Healthy life expectancy,Freedom to make life choices,Generosity,Perceptions of corruption,GDP_Level
153,154,Afghanistan,3.203,0.350,0.517,0.361,0.000,0.158,0.025,Low
154,155,Central African Republic,3.083,0.026,0.000,0.105,0.225,0.235,0.035,Low
155,156,South Sudan,2.853,0.306,0.575,0.295,0.010,0.202,0.091,Low


## 10. Key Insights

**Based on the EDA using Plotly visualizations, here are the main findings:**

1. **Clean Data**: The dataset has 0% missing values, making it ideal for modeling without any imputation required.
2. **Top Happiness Drivers**: The correlation heatmap shows that `GDP per capita` (0.79), `Social support` (0.78), and `Healthy life expectancy` (0.78) are the strongest predictors of a country's Happiness Score.
3. **Outliers**: Features like `Generosity` and `Perceptions of corruption` are highly right-skewed. Only a few specific countries act as extreme positive outliers, demonstrating exceptionally high levels in these areas compared to the global average.
4. **Economic Impact (GDP_Level)**: Boxplot analysis reveals a stark contrast between economic groups. Countries in the "High" GDP tier have a median Happiness Score over 6.5, whereas those in the "Low" tier struggle to reach 4.5.
